plugin/forward/persistent.go

```diff
@@ -96,13 +92,49 @@ Wait:
 
 			t.conns["tcp-tls"] = append(t.conns["tcp-tls"], &persistConn{conn, time.Now()})
 
+		case <-ticker.C:
+			t.cleanup(false)
+
 		case <-t.stop:
+			t.cleanup(true)
 			close(t.ret)
 			return
 		}
 	}
 }
 
+// closeConns closes connections.
+func closeConns(conns []*persistConn) {
+	for _, pc := range conns {
```

Comment:
I remove these connections from `transport.conns` before calling `closeConns()`, so they are unreachable from transport anymore

Reply:
But `conns` is and because we're only `ranging` over it; that makes it save?

Reply:
`conns` here is just a parameter passed to `closeConns()`, i.e. local variable. This is not the same as `transport.conns`

Reply:
it's a slice, so the underlaying array is shared between the two; it still don't think this is safe to do...

Reply:
yes, the `conns` is a slice, but the underlaying array is not shared. The second reference to this array is erased from `transport.conns` right before calling `closeConns()`, see the lines
https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR71
and
https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR133

And seems I missed the line
https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR122
which also should be moved before `closeConns()`

We actually don't expire and search connections in a separate goroutine. The `cleanup()` is called synchronously from `connManager()`. We only call  `closeConns()` in a separate goroutine which just closes connections which are not accessible from `connManager()` anymore

Reply:
Well, strictly speaking, in case of https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR133 the same underlying array is still used, but slices are not intersected. I.e. `connManager()` has a slice with no references to the connections passed to `closeConns()` with another slice. Adding new connection to cache cannot affect the connections passed to `closeConns()`

Reply:
ok

Reply:
I don't think that's true. I *think* the little program below will simulate this situation:

```
package main

import (
        "log"
        "time"
)

func close(s []int) {
        log.Println("Pre-sleep closing: ", s)
        time.Sleep(1*time.Second)
        log.Println("Post-sleep closing: ", s)
}

func add(s []int) {
}

func main() {
        log.Println("Hi")

        s := make([]int,10)
        for i := 0; i < 10; i++ {
                s[i] = i
        }

        ss := s[5:]
        s = s[:5]
        go close(ss)

        time.Sleep(time.Second/2)
        s = append(s,100)

        time.Sleep(2*time.Second)
        log.Println("s = ", s)
}
```

Running this we see:

```
jbelamaric@ANN-M-JBELAMARIC [k8s:minikube] ~/slices$ go run main.go
2018/05/23 18:28:30 Hi
2018/05/23 18:28:30 Pre-sleep closing:  [5 6 7 8 9]
2018/05/23 18:28:31 Post-sleep closing:  [100 6 7 8 9]
2018/05/23 18:28:32 s =  [0 1 2 3 4 100]
jbelamaric@ANN-M-JBELAMARIC [k8s:minikube] ~/slices$
```

Is there something different about that and the situation here?

Reply:
@johnbelamaric , your case is different

I pass to the `closeConns()` goroutine the slice which points to the beginning part of the array. And the slice pointing to remaining part overwrites the original slice in transport object which previously pointed to whole array . So, the transport cannot access the elements which resided in the beginning of the original array anymore. If transport appends new element it will be either added to the end of original array, or added to a newly allocated array after copying elements which were accessible from the transport slice. In any case the elements passed to `closeConns()` will remain in place and will not change.

To simulate this situaltion, you need to replace
```
        ss := s[5:]
        s = s[:5]
```
in your example with 
```
        ss := s[:5]
        s = s[5:]
```

Reply:
Ah, got it. Ok 👍

Reply:
Thanks both.

We're good to go then?

Reply:
done